[Check-in][SP-guide] Add examples using PKCE (#332)
EGI-Federation/documentation: content/en/providers/check-in/sp/_index.md

This page contains information about connecting services to
[EGI Check-in](https://www.egi.eu/services/check-in/) in order to allow user
login through Check-in and to receive users' attributes. Check-in is connected
to a wide range of academic and social Identity Providers that users can choose
from in order to access your service.

## Services eligible for integration

EGI Operations, as owner of the Check-in service, must approve every request for

> integration of new services with Check-in. The approval (or non-approval) will

Reviewer: ```suggestion
integration of new services with Check-in. The approval (or non-approval) is
```

Author: Resolved by commit [Fix typos](https://github.com/EGI-Federation/documentation/pull/332/commits/7c9c7855873f8858ed9f14a4a5119ff667c32741)

> be based on some pre-requisites, the relevance of the service for the EGI

Reviewer: ```suggestion
based on some prerequisites, the relevance of the service for the EGI
```

Author: Resolved by commit [Fix typos](https://github.com/EGI-Federation/documentation/pull/332/commits/7c9c7855873f8858ed9f14a4a5119ff667c32741)

   instance of EGI Check-in by selecting the "Demo" integration environment
   during registration through the EGI Federation Registry Portal. The demo
   instance allows for testing authentication and authorisation through the
   academic and social Identity Providers connected to Check-in without
   affecting the production Check-in service. Note that while the demo instance
   has identical functionality to the production instance, no information is
   shared between the two systems.
   - You can also test new features of Check-in that are not available in
     production yet, by registering your Service Provider and testing
     integration with the **development** instance of Check-in. In the
     development instance service requests can be self-reviewed without the need
     to wait for approval from an administrator. As with the demo instance, the
     development instance allows for testing authentication and authorisation
     without affecting the production Check-in service. **NB: the list of
     supported Identity Providers in the development instance is limited.
     Therefore, we recommend using the [EGI SSO](https://sso.egi.eu/admin/) or

>      any of the social identity providers to test the login workflow when using

Reviewer: I would revert the order: using social accounts or the EGI SSO. We don't really want more people to create accounts in the EGI SSO unless they have no other choice.

Author: Resolved by commit [Fix typos](https://github.com/EGI-Federation/documentation/pull/332/commits/7c9c7855873f8858ed9f14a4a5119ff667c32741)

Example request:

```shell
curl -X POST "${TOKEN_ENDPOINT}" \

> -d "grant_type=authorization_code" \

Reviewer: ```suggestion
  -d "grant_type=authorization_code" \
```

Author: Resolved by commit [Add indent to curl data parameter](https://github.com/EGI-Federation/documentation/pull/332/commits/d4c45e1af19edd9a4d4cdaa64e897a218160bc87)

> -d "code=${CODE}" \
> -d "client_id=${CLIENT_ID}" \
> -d "redirect_uri=${REDIRECT_URI}" \
> -d "code_verifier=${CODE_VERIFIER}" | python -m json.tool

Reviewer: ```suggestion
  -d "code=${CODE}" \
  -d "client_id=${CLIENT_ID}" \
  -d "redirect_uri=${REDIRECT_URI}" \
  -d "code_verifier=${CODE_VERIFIER}" | python -m json.tool
```

Author: Resolved by commit [Add indent to curl data parameter](https://github.com/EGI-Federation/documentation/pull/332/commits/d4c45e1af19edd9a4d4cdaa64e897a218160bc87)

Example request:

```shell
curl -X POST "${TOKEN_ENDPOINT}" \

> -u "${CLIENT_ID}":"${CLIENT_SECRET}" \

Reviewer: ```suggestion
  -u "${CLIENT_ID}":"${CLIENT_SECRET}" \
```

Author: Resolved by commit [Add indent to curl data parameter](https://github.com/EGI-Federation/documentation/pull/332/commits/d4c45e1af19edd9a4d4cdaa64e897a218160bc87)

-d "grant_type=refresh_token" \

> -d "refresh_token=${REFRESH_TOKEN}" \
> -d "scope=openid%20email%20profile" | python -m json.tool;

Reviewer: ```suggestion
  -d "refresh_token=${REFRESH_TOKEN}" \
  -d "scope=openid%20email%20profile" | python -m json.tool;
```

Author: Resolved by commit [Add indent to curl data parameter](https://github.com/EGI-Federation/documentation/pull/332/commits/d4c45e1af19edd9a4d4cdaa64e897a218160bc87)

}
```

###### Refresh Request with PKCE

To combine the refresh token grant type with PKCE you need to make the following
request:

```shell
curl -X POST "${TOKEN_ENDPOINT}" \

> -d "client_id=${CLIENT_ID}" \
> -d "grant_type=refresh_token" \
> -d "refresh_token=${REFRESH_TOKEN}" \
> -d "scope=openid%20email%20profile" | python -m json.tool;

Reviewer: ```suggestion
  -d "client_id=${CLIENT_ID}" \
  -d "grant_type=refresh_token" \
  -d "refresh_token=${REFRESH_TOKEN}" \
  -d "scope=openid%20email%20profile" | python -m json.tool;
```

Reviewer: Please do the same for other code excerpts

Author: Resolved by commit [Add indent to curl data parameter](https://github.com/EGI-Federation/documentation/pull/332/commits/d4c45e1af19edd9a4d4cdaa64e897a218160bc87)

### Example OIDC Client

In this guide we will demonstrate how to install and configure a
[Simple OIDC Client](https://github.com/rciam/simple-oidc-client-php).

#### Install simple-oidc-client-php

This guide assumes the Apache HTTP server has been installed and the document
root is `/var/www/html`

Move to the apache document root and download and extract
[simple-oidc-client-php-v2.0.0.zip](https://github.com/rciam/simple-oidc-client-php/releases/download/v2.0.0/simple-oidc-client-php-v2.0.0.zip).

#### Configure Client

> Go to this link and login <https://aai.egi.eu/federation>

Reviewer: ```suggestion
Login to the [EGI Federation Registry](https://aai.egi.eu/federation)
```

Author: Resolved by commit [Fix typos](https://github.com/EGI-Federation/documentation/pull/332/commits/7c9c7855873f8858ed9f14a4a5119ff667c32741)

Then create a new service or edit your existing service. In `General` tab fill
all the required fields. For `Integration Environment` select `Demo`. In
`Protocol Specific` tab select as Protocol the `OIDC Service` and then in the
`Redirect URI(s)` insert your simple-oidc-client-php URL (e.g.

> <http://localhost/simple-oidc-client-php/refreshtoken.php>). This URL must link

Reviewer: ```suggestion
`http://localhost/simple-oidc-client-php/refreshtoken.php`). This URL must link
```

Author: Resolved by commit [Fix typos](https://github.com/EGI-Federation/documentation/pull/332/commits/7c9c7855873f8858ed9f14a4a5119ff667c32741)

to `refreshtoken.php` which is located in simple-oidc-client-php directory.
Next, in `Scope` select the scopes that your service needs. Then, submit the
form and and self approve it. Finally you should get a pair of `Client ID` and
`Client Secret`.

#### Configure simple-oidc-client-php

Now that you have everything you need, you can configure your login settings. Go
to your terminal and open `config.php` with your favorite text editor.

Example:

```shell
vi simple-oidc-client-php/config.php
```

Let's go quickly through the settings:

> - `title` required, is the title on the navigation bar
> - `img` required, is the source of the logo
> - `scope_info` optional, is a message that informs the user for the application
>   requirements
> - `issuer` required, is the base URL of our IdentityServer instance. This will
>   allow oidc-client to query the metadata endpoint so it can validate the tokens
> - `client_id` required, is the id of the client we want to use when hitting the
>   authorization endpoint
> - `client_secret` optional, a value the offers better security to the message
>   flow
> - `pkceCodeChallengeMethod` optional, a string that defines the code challenge
>   method for PKCE. Choose between `plain` or `S256`.
> - `redirect_url` required, is the redirect URL where the client and the browser
>   agree to send and receive correspondingly the code
> - `scopesDefine` required, defines the scopes the client supports
> - `refresh_token_note` optional, info for the refresh token
> - `access_token_note` optional, info for the access token
> - `manage_token_note` optional, message the informs the user where can manage
>   his tokens
> - `manageTokens` optional, URL of the manage tokens service
> - `sessionName` required, define the name of the cookie session
> - `sessionLifetime` required, define the duration of the session. This must be
>   equal to the validity time of the access token.

Reviewer: ```suggestion
- `title` required, the title on the navigation bar
- `img` required, the source of the logo
- `scope_info` optional, a message that informs the user for the application
  requirements
- `issuer` required, the base URL of our IdentityServer instance. This will
  allow oidc-client to query the metadata endpoint so it can validate the tokens
- `client_id` required, the id of the client we want to use when hitting the
  authorization endpoint
- `client_secret` optional, a value the offers better security to the message
  flow
- `pkceCodeChallengeMethod` optional, a string that defines the code challenge
  method for PKCE. Choose between `plain` or `S256`.
- `redirect_url` required, the redirect URL where the client and the browser
  agree to send and receive correspondingly the code
- `scopesDefine` required, defines the scopes the client supports
- `refresh_token_note` optional, info for the refresh token
- `access_token_note` optional, info for the access token
- `manage_token_note` optional, message the informs the user where can manage
  his tokens
- `manageTokens` optional, URL of the manage tokens service
- `sessionName` required, define the name of the cookie session
- `sessionLifetime` required, define the duration of the session. This must be
  equal to the validity time of the access token.
```

Author: Resolved by commit [Fix typos](https://github.com/EGI-Federation/documentation/pull/332/commits/7c9c7855873f8858ed9f14a4a5119ff667c32741)
